doc/dev/architecture.md

```diff
@@ -0,0 +1,304 @@
+# Architecture
+
```

Comment:
Is there another `architecture.rst` file in the `docs` folder?

Reply:
I'm writing the document in a separate markdown file without touching the original `architecture.rst` yet, because this way I can actually preview it properly.

Reply:
Text has been moved to the markdown file: https://github.com/PennyLaneAI/catalyst/pull/160/commits/59cb043edb7e7118a2ee0218bf3e1e07eb0ab2b6